Floofbot/Services/RaidProtectionService.cs

```diff
@@ -0,0 +1,387 @@
+using System.Diagnostics.Eventing.Reader;
+using System.Linq;
+using System.Net.NetworkInformation;
+using System.Security.Policy;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+namespace Floofbot.Services
+{
+    public class RaidProtectionService
+    {
+        // will store user id and how many counts they've had
+        private Dictionary<ulong, int> userPunishmentCount = new Dictionary<ulong, int>();
+        // used to keep track of the number of messages a user sent for spam protection
+        private Dictionary<ulong, int> userMessageCount = new Dictionary<ulong, int>();
+        // a list of punished users used to detect any potential raids
+        private List<SocketUser> punishedUsers = new List<SocketUser>();  
+        // determines how long before a user is forgiven for a punishment
+        private static int forgivenDuration = 5 * 60 * 1000; // 5 min
+        // determines the rate at which users can send messages, currently no more than x messages in y seconds
+        private static int durationForMaxMessages = 5 * 1000; // 5 s
+        // determines the max number of punishments a user can have before being punished
+        private static int maxNumberOfPunishments = 3;
+        // the delay before the bot msg is deleted in ms
+        private static int botMessageDeletionDelay = 3000;
+        // the duration before a user is removed from the list of punished users
+        private static int removePunishedUserDelay = 30 * 60 * 1000; // 30 min
+        // the number of punished users within a time frame before the mods are alerted of possible raids
+        private static int maxNumberPunishedUsers = 3;
+        // These are used to determine if there are an excessive number of joins in a short time frame
+        private static int maxNumberOfJoins = 5;
+        private static int userJoinsDelay = 2 * 60 * 1000; // 2 min
+        private Dictionary<IGuild, int> numberOfJoins = new Dictionary<IGuild, int>();
+        // The max number of repeated emojis before triggering the raid protection
+        private static int maxNumberEmojis = 5;
+
+
+
+        public RaidProtectionService()
+        {
+
+        }
+        public RaidProtectionConfig GetServerConfig(IGuild guild, FloofDataContext _floofDb)
+        {
+            RaidProtectionConfig serverConfig = _floofDb.RaidProtectionConfigs.Find(guild.Id);
+            return serverConfig;
+        }
+        public async Task NotifyModerators(SocketRole modRole, ITextChannel modChannel, string message)
+        {
+            await modChannel.SendMessageAsync(modRole.Mention + " there may be a possible raid! Reason: ``" + message + "``");
+        }
+        private async void UserPunishmentTimeout(ulong userid)
+        {
+            await Task.Delay(forgivenDuration);
+            if (userPunishmentCount.ContainsKey(userid))
+                if (userPunishmentCount[userid] != 0)
```

Comment:
These `if`s can be evaluated on the same line

Reply:
Same thing occurs twice below

Reply:
If I ran these on the same line the second if statement will return null if they key does not exist

Reply:
I'm not sure if I follow what you are trying to say, but
`if (userPunishmentCount.ContainsKey(userid) && userPunishmentCount[userid] != 0)` should work, because if `userPunishmentCount.ContainsKey(userid)` evaluates to `false` (I'm assuming `null == false`), the second statement is not evaluated.